addons/product/models/product_attribute.py

```diff
@@ -447,7 +447,14 @@ def write(self, values):
                         _("You cannot change the product of the value %s set on product %s.") %
                         (ptav.display_name, ptav.product_tmpl_id.display_name)
                     )
-        return super(ProductTemplateAttributeValue, self).write(values)
+        res = super(ProductTemplateAttributeValue, self).write(values)
+
+        exclude_in_value = 'exclude_for' in values
+        if exclude_in_value:
+            self._unarchive_variants()
+            self._archive_variants_with_exclusion()
```

Comment:
why two separated methods? if fused both into one, then you could archive the same ones that were already archived in the beginning and the method would seem safer. I mean, one user maybe want to have archived a variant although the variant is theoretically possible.

Reply:
Yep, it does not make sense at this stage. At the beginning there were two paths in which we could discriminate and we only archive or unarchive and archive.